### ISSUE TITLE
`convex_hull_image` returns different results between 0.18.0 and 0.19.3

### ISSUE THREAD

**Issue by dudulasry** [scikit-image/scikit-image#6510]
## Description

Hi guys, my team and I recently decided to migrate to `scikit-image==0.19.3` due to a bigger migration to Python 3.10. Our algorithms rely heavily on `scikit-image` (0.14.2).

During the migration process, we encountered a difference in the `convex_hull_image` function. Seems like the underlying calculation was changed. I found this [issue](https://github.com/scikit-image/scikit-image/issues/3892) ([relevant PR](https://github.com/scikit-image/scikit-image/pull/5029)), and I saw that in the newer versions, the condition that checks whether a point is in a polygon was changed from this:

<b> Pre 0.18.1 </b>
```
min <= point < max
```

<b> Post 0.18.1 </b>
```
min <= point <= max
```

This is a bit problematic for us, because it changes the results of our algorithms pretty dramatically. I tried to look at the code a bit and I saw that `point_in_polygon` now returns 4 possible values: 
```
0: outside, 1: inside, 2: vertex; 3: edge
``` 

However, the function that `convex_hull_image` uses casts the array of numbers into a boolean mask. So for an array like this:
```
[0, 1, 2, 3]
```
It'd return this:
```
[False, True, True, True]
```

This implicitly includes edge/vertex points inside the polygon. This is a problem because our algorithms were not trained with this kind of shapes.

Is there a way to change this behavior outside of `scikit-image`? I couldn't find any flag that controls this behavior. The most feasible option that we thought about is to fork `scikit-image` and modify the `convex_hull_image` -> `grid_points_in_poly` -> `_grid_points_in_poly` to not cast the labels output to boolean and instead control this behavior with a flag. 

However, we prefer not to fork this project and change its internals.
Any suggestions would be highly appreciated!

## Way to reproduce
Run `convex_hull_image` in 0.14.2 and in 0.19.3 and witness the results. The mask in 0.19.3 will be slightly larger, due to the inclusion of edge/vertex points.

## Version information
```python
# Paste the output of the following python commands
from __future__ import print_function
import sys; print(sys.version)
import platform; print(platform.platform())
import skimage; print(f'scikit-image version: {skimage.__version__}')
import numpy; print(f'numpy version: {numpy.__version__}')
```

#### Output
```python
3.10.5 (main, Aug 14 2022, 07:33:04) [GCC 9.4.0]
Linux-5.15.0-1015-aws-x86_64-with-glibc2.31
scikit-image version: 0.19.3
numpy version: 1.23.1
```

**Comment by grlee77** [comment]
Thanks for reporting this @dudlasry. 

> The most feasible option that we thought about is to fork scikit-image and modify the convex_hull_image -> grid_points_in_poly -> _grid_points_in_poly to not cast the labels output to boolean and instead control this behavior with a flag.

Having a flag to offer finer control of the behavior seems like it could be a reasonable enhancement here. I am not too familiar with this issue, though. Perhaps @rfezzani can offer some perspective as it seems a [more general user choice was mentioned previously as an option](https://github.com/scikit-image/scikit-image/issues/3892#issuecomment-494296268).

**Comment by rfezzani** [comment]
Thank you @dudulasry for the report and sorry for the delayed answer :pray: 

You in fact identified the issue/PR that caused this behavior change. The point is that even if you consider edges and corners as being not included in polygons, you will not retrieve exactly the same behavior as v0.14 because that behavior was inconsistent, as it was reported in #3892 :confused:
If you want to try anyway removing edges and corners from inside the polygon, the solution would be to
- make `skimage.measure._pnpoly._grid_points_in_poly` return the label of the points instead of casting it to bool ([here](https://github.com/scikit-image/scikit-image/blob/80c98ce22faf33418b8493d3ebbe4605e47f11eb/skimage/measure/_pnpoly.pyx#L55))
- in `skimage.morphology.convex_hull_image` define the mask as `mask = grid_points_in_poly(image.shape, vertices) == 1`
[here](https://github.com/scikit-image/scikit-image/blob/80c98ce22faf33418b8493d3ebbe4605e47f11eb/skimage/morphology/convex_hull.py#L155)

**Comment by dudulasry** [comment]
@rfezzani Thanks for the answer.
This is kind of the change I implemented in this pull request #6515. I think that users of `scikit-image` should be able to decide whether they want to include vertex/edge points in their final hull mask.

Is there a chance #6515 will be merged to the main branch and be part of the next release?

**Comment by grlee77** [comment]
> Is there a chance https://github.com/scikit-image/scikit-image/pull/6515 will be merged to the main branch and be part of the next release?

I think we can get it in for the next release given that it seems to already be nearly ready to go. We don't have a specific date for the release, but I think likely not far in the future as Python 3.11 is being released next month and we will want to have a version providing wheels for that. I just updated the wheel building and tests for our PyWavelets dependency yesterday and will release it this week. We can then try out the same changes here. We use `cibuildwheel` for building wheels and it already supports building wheels with Python 3.11-rc1 (guaranteed to be binary compatible with the final Python 3.11 release, so we don't necessarily have to wait until the final release to provide compatible wheels).

**Comment by dudulasry** [comment]
Hi guys, I now I understand that even after filtering all the points which don't have label == 1, I still get a different mask between 0.14.2 and 0.19.3. I expected them to be the same.

Do you know how can I obtain the original masks, given the latest code? Is there any manipulation I can do to the input so that the output would be identical to 0.14.2?

**Comment by SharonSv** [comment]
> Hi guys, I now I understand that even after filtering all the points which don't have label == 1, I still get a different mask between 0.14.2 and 0.19.3. I expected them to be the same. @rfezzani I think I missed the part in your comment that mentioned it:
> 
> > The point is that even if you consider edges and corners as being not included in polygons, you will not retrieve exactly the same behavior as v0.14 because that behavior was inconsistent, as it was reported in #3892 😕
> 
> Do you know how can I obtain the original masks, given the latest code? Is there any manipulation I can do to the input so that the output would be identical to 0.14.2?

Hi all, 
I am working with @dudulasry and it is a very challenging point for us, we will appreciate any help here.

As @dudulasry mentioned, adding this flag won't help to solve our issue and the received mask - with or without the edges - is still different from what we get if we use 0.18.0 as the logic itself changed and the edges identified in 0.18.0 are different from the edges identified in 0.19.3.

This change is critical and not backward compatible.

I am suggesting is to add a `mode` kwarg which can be one of two:
1. The logic after 0.18.0 (default)
2. The logic before 0.18.0

I am not sure about the options naming

@rfezzani, @grlee77 WDYT?
Thanks, and as I said, we will appreciate any help and any other suggestion to overcome this challenge

**Comment by rfezzani** [comment]
The behavior change you observe is resulting from a bug fix (#3892). This legitimated breaking the backward compatibility :confused:.
Switching to python 3.10 prevents you from installing `scikit-image` v0.18 from regular pip or conda repositories. Can't you compile and install `skimage` locally?
An other solution would be to install skimage from github directly using pip:
```
pip install git+https://github.com/scikit-image/scikit-image.git@v0.18.3
```
But this actually doesn't work because of
https://github.com/scikit-image/scikit-image/blob/79cb22aef09dc8c03cb60f1f9d74e11a71abc17f/.gitmodules#L3

This is fixed in v0.19 (#6283) but not for v0.18 :sweat:. We investigate...
